### PR TITLE
[ONNX] Add constant fold for onnx::ReduceProd

### LINKED ISSUE #75621: [ONNX] Export embedding_bag with 2d input

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #74082
* #74081

One of the origins for `onnx::ReduceProd` is `aten::numel`.
Adding constant fold support for `onnx::ReduceProd` closes the gap
and enables symbolic shape inference for `aten::numel` nodes that
has static shape input.

One example is `torch.nn.EmbeddingBag` when input is 2d. An `Offset`
tensor will be created by `tensor.numel()`. This `Offset` can be
properly exported as constant now, if the input has static shape.
Otherwise, `Offset` will be considered to have dynamic value, and
will not be supported in opset 10. Support for dynamic `Offset` 
requires opset 11. 

Fixes: #75621

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -6486,9 +6486,6 @@ def test_embedding_bag(self):
         offset = torch.tensor([0, 2, 5, 6])
         self.run_test(model, (input, offset))
 
-    @disableScriptTest()  # error in propagate as assign input shape
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:6489]: Can you please update the PR description with more details?
In particular, I can see why constant folding is an improvement (we should do it whenever possible), but I would imagine any ops that we don't support for constant folding should not cause failures. Is there still a bug with handling shape propagation of `ReduceProd` nodes? Will we trigger that failure if we disable constant folding when exporting? If not, why not?

Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:6489]: appended with a line of comment in description.

> Is there still a bug with handling shape propagation of ReduceProd nodes?

The "shape" that gets used later on in this use case, is the value data in ReduceProd nodes. So normal shape propagation wouldn't help. We need to compute ReduceProd, and store the value in `ConstantValueMap::ShapeValue`, for further "symbolic" shape inference later.

> Will we trigger that failure if we disable constant folding when exporting?

No, it is symbolic shape inference that utilizes part of the constant folding code to compute some shape value. It is unrelated to the `do_constant_folding` feature that folds model parameters.

Comment (garymm) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:6489]: Can you add these details to the PR description?

**Review by garymm — APPROVED** [review]

**Pull request closed without merging**